File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(defn is-upper [message]
  (and
    (= (str/upper-case message) message)
    (not= (str/lower-case message) message)))


(defn is-question [message]
  (str/ends-with? message "?"))
  

(defn response-for
  "Bob's responses to different statement"
  ([message]
   (cond
     (str/blank? message) "Fine. Be that way!"
     (and (is-upper message) (is-question message)) "Calm down, I know what I'm doing!"
     (is-upper message)   "Whoa, chill out!"
     (is-question message) "Sure."
     :else "Whatever.")))  

File: elixir/bob/lib/bob.ex
defmodule Bob do
  def hey(input) do
    cond do
      upper?(input) && question?(input) && alpha?(input) ->
        "Calm down, I know what I'm doing!"

      empty?(input) ->
        "Fine. Be that way!"

      upper?(input) && alpha?(input) ->
        "Whoa, chill out!"

      question?(input) ->
        "Sure."

      true ->
        "Whatever."
    end
  end

  defp alpha?(input), do: String.upcase(input) != String.downcase(input)
  defp empty?(input), do: String.trim(input) == ""
  defp question?(input), do: input |> String.trim() |> String.last() == "?"
  defp upper?(input), do: String.upcase(input) == input
end
